packages/gensx/src/checkpoint.ts

```diff
@@ -199,6 +220,217 @@ export class CheckpointManager implements CheckpointWriter {
+      // Initialize secrets set for this node
+      let nodeSecrets = this._secretValues.get(nodeId);
+      if (!nodeSecrets) {
+        nodeSecrets = new Set();
+        this._secretValues.set(nodeId, nodeSecrets);
+      }
+
+      // Use paths purely for collection
+      for (const path of paths) {
+        const value = this.getValueAtPath(props, path);
+        if (value !== undefined) {
+          this.collectSecretValues(value, nodeSecrets);
+        }
+      }
+    });
+  }
+
+  private collectSecretValues(
+    data: unknown,
+    nodeSecrets: Set<unknown>,
+    visited = new WeakSet(),
+  ): void {
+    // Skip if already visited to prevent cycles
+    if (data && typeof data === "object") {
+      if (visited.has(data)) {
+        return;
+      }
+      visited.add(data);
+    }
+
+    // Handle primitive values
+    if (typeof data === "string") {
+      if (data.length >= this.MIN_SECRET_LENGTH) {
+        nodeSecrets.add(data);
+      }
+      return;
+    }
+
+    // Skip other primitives
+    if (!data || typeof data !== "object") {
+      return;
+    }
+
+    // Handle arrays and objects (excluding ArrayBuffer views)
+    if (Array.isArray(data) || !ArrayBuffer.isView(data)) {
+      const values = Array.isArray(data) ? data : Object.values(data);
+      values.forEach(value => {
+        this.collectSecretValues(value, nodeSecrets, visited);
+      });
+    }
+  }
+
+  private scrubSecrets(data: unknown, nodeId?: string, path = ""): unknown {
+    return this.withNode(nodeId ?? "", () => {
+      // Handle primitive values
+      if (typeof data === "string" || typeof data === "number") {
+        const strValue = String(data);
+        return this.scrubString(strValue);
+      }
+
+      // Skip other primitives
+      if (!data || typeof data !== "object") {
+        return data;
+      }
+
+      // Handle arrays
+      if (Array.isArray(data)) {
+        return data.map((item, index) =>
+          this.scrubSecrets(
+            item,
+            nodeId,
+            path ? `${path}.${index}` : `${index}`,
+          ),
+        );
+      }
+
+      // Handle objects (excluding ArrayBuffer views)
+      if (!ArrayBuffer.isView(data)) {
+        return Object.fromEntries(
+          Object.entries(data).map(([key, value]) => [
+            key,
+            this.scrubSecrets(value, nodeId, path ? `${path}.${key}` : key),
+          ]),
+        );
+      }
+
+      return data;
+    });
+  }
+
+  private scrubString(value: string): string {
+    const effectiveSecrets = this.getEffectiveSecrets();
+    let result = value;
+
+    // Sort secrets by length (longest first) to handle overlapping secrets correctly
+    const secrets = Array.from(effectiveSecrets)
+      .filter(s => typeof s === "string" && s.length >= this.MIN_SECRET_LENGTH)
+      .sort((a, b) => String(b).length - String(a).length);
+
+    // Replace each secret with [secret]
+    for (const secret of secrets) {
+      if (typeof secret === "string") {
```

Comment:
how do we get secrets that aren't strings?

Reply:
`scrubSecrets` is a recursive function that processes arrays and objects. When it hits a number or string, it tostrings it and calls this function. 

```ts
      if (typeof data === "string" || typeof data === "number") {
        const strValue = String(data);
        return this.scrubString(strValue);
      }
```